sympy/integrals/integrals.py

```diff
@@ -1242,6 +1242,118 @@ def _sage_(self):
+        on the real axis.
+        In mathematics, the Cauchy principal value, is a method for assigning values to certain improper
+        integrals which would otherwise be undefined.
+
+        Examples
+        ========
+
+        >>> from sympy import Dummy, symbols, integrate, limit, oo
+        >>> from sympy.integrals.integrals import Integral
+        >>> from sympy.calculus.singularities import singularities
+        >>> x = symbols('x')
+
+        >>> g = x + 1
+        >>> Integral(g, (x, -oo, oo)).principal_value()
+        oo
+
+        >>> f = 1/(x**3)
+        >>> singularities(f,x)
+        {0}
+
+        >>> Integral(f, (x, -oo, oo)).principal_value()
+        0
+
+        >>> Integral(f, (x, -oo, 0)).principal_value()
+        -oo
+
+        >>> Integral(f, (x, 0, oo)).principal_value()
+        oo
+
+        >>> Integral(f, (x, 1, 2)).principal_value()
+        3/8
+
+        >>> Integral(f, (x, -2, -1)).principal_value() + Integral(f, (x, 1, 2)).principal_value()
+        0
+
+        >>> Integral(f, (x, 0.4, 3)).principal_value()
+        3.06944444444444
+
+        >>> Integral(f, (x, -3, -0.4)).principal_value() + Integral(f, (x, 0.4, 3)).principal_value()
+        0
+
+
+        References
+        ==========
+
+        .. [1] http://en.wikipedia.org/wiki/Cauchy_principal_value
+        .. [2] http://mathworld.wolfram.com/CauchyPrincipalValue.html
+        """
+        from sympy.calculus import singularities
+
+        if len(self.limits) != 1 or len(list(self.limits[0])) != 3:
+            raise ValueError("You need to insert a variable, lower_limit, and upper_limit correctly to calculate "
+                             "cauchy's principal value")
+
+        x, a, b = self.limits[0]
+
+        if not(a.is_comparable and b.is_comparable and a <= b):
+            raise ValueError("The lower_limit must be smaller than or equal to the upper_limit to calculate "
+                             "cauchy's principal value. Also, a and b need to be comparable.")
+
+        if a == b:
+            return 0
+        r = Dummy('r')
+        n = Dummy('n')
+
+        singularities_list = [s for s in singularities(self.function, x) if s.is_comparable and a < s < b]
+
+        f = self.function
+        F = integrate(f, x)
+
+        if F.has(Integral):
+            return self
+        if len(singularities_list) == 0:
+            if a is -oo and b is oo:
+                I = limit(F - F.subs(x, -x), x, oo)
+                return I
+            else:
+                I = limit(F, x, b, '-') - limit(F, x, a, '+')
+                return I
+
+        if len(singularities_list) == 1:
+            I = limit(((F.subs(x, singularities_list[0] - r)) - F.subs(x, singularities_list[0] + r)), r, 0,
+                      '+') - limit(F, x, a).subs([(oo, n), (-oo, -n)]) + limit(F, x, b).subs([(oo, n), (-oo, -n)])
+            return I.subs([(n, oo), (-n, -oo)])
```

Comment:
What is the total effect of the two substitutions: first `n` for `oo` and then again `oo` for `n`?

Reply:
@jksuom  Hope you understand what I'm trying to say.

Reply:
Can you give an example where this repeated substitution would be useful?

Reply:
@jksuom  Ofcourse 
```
avi@avi-Aspire-A515-51G:~/sympy$ python3
Python 3.6.5 |Anaconda, Inc.| (default, Apr 29 2018, 16:14:56) 
[GCC 7.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import Dummy, symbols, integrate, limit, oo, sqrt
>>> from sympy.integrals.integrals import Integral
>>> x = symbols('x')
>>> r = Dummy('r')
>>> n = Dummy('n')
>>> f = (x**2 - 4)*(x + 2)/(x-1)
>>> F = integrate(f,x)
>>> F
x**3/3 + 3*x**2/2 - x - 9*log(x - 1)
>>> from sympy.calculus.singularities import singularities
>>> singularities_list = [s for s in singularities(f, x)]
>>> singularities_list 
[1]
>>> a = -oo
>>> b = oo
>>> I = limit(((F.subs(x, singularities_list[0] - r)) - F.subs(x, singularities_list[0] + r)), r, 0,'+') - limit(F, x, a).subs([(oo, n), (-oo, -n)]) + limit(F, x, b).subs([(oo, n), (-oo, -n)])
>>> I
2*_n - 9*I*pi
>>> I.subs([(n, oo), (-n, -oo)])
oo - 9*I*pi

```

Reply:
@jksuom  Hope this satisfies your query.

Reply:
What will happen if these substitutions are removed?

Reply:
@jksuom The answer will come in terms of `_n` (if the last substitution is removed)
While if all of the `subs` are removed, then integrals of functions like `g = 1/x` will come out as `nan` again which is wrong.(The correct answer should be `0` ==> which is coming with using the substitutions)

Why do you think there is an issue with using substitution? Because I don't find anything wrong with it. Also, I've used them carefully so that there's no wastage of a substitution where it isn't required.

Reply:
@jksuom Okay, sure
```
avi@avi-Aspire-A515-51G:~/sympy$ python3
Python 3.6.5 |Anaconda, Inc.| (default, Apr 29 2018, 16:14:56) 
[GCC 7.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import Dummy, symbols, integrate, limit, oo, sqrt
>>> from sympy.integrals.integrals import Integral
>>> r = Dummy('r')
>>> n = Dummy('n')
>>> x = symbols('x')
>>> a = -oo
>>> b = oo
>>> from sympy.calculus.singularities import singularities
>>> f = 1/x
>>> F = integrate(f,x)
>>> singularities_list = [s for s in singularities(f, x)]
>>> I = limit(((F.subs(x, singularities_list[0] - r)) - F.subs(x, singularities_list[0] + r)), r, 0,'+') - limit(F, x, a) + limit(F, x, b)
>>> I
nan
>>> I2 = limit(((F.subs(x, singularities_list[0] - r)) - F.subs(x, singularities_list[0] + r)), r, 0,'+') - limit(F, x, a).subs([(oo, n), (-oo, -n)]) + limit(F, x, b).subs([(oo, n), (-oo, -n)])
>>> I2
I*pi
```

Reply:
@jksuom 
Here, `I` doesn't use substitution
```
>>> I = limit(((F.subs(x, singularities_list[0] - r)) - F.subs(x, singularities_list[0] + r)), r, 0,'+') - limit(F, x, a) + limit(F, x, b)
>>> I
nan
```
while `I2` does use substitution
```
>>> I2 = limit(((F.subs(x, singularities_list[0] - r)) - F.subs(x, singularities_list[0] + r)), r, 0,'+') - limit(F, x, a).subs([(oo, n), (-oo, -n)]) + limit(F, x, b).subs([(oo, n), (-oo, -n)])
>>> I2
I*pi
```

Reply:
If the limits are `-oo, oo`, then the difference `- limit(F, x, a) + limit(F, x, b)` is wrong. There should be only one limit as on line 1322, and there is no need for substitutions `n -> oo` or `oo -> n`. In the remaining cases the difference on line 1325 shall be computed (and no similar substitutions are needed, either).

Reply:
@jksuom 
No sir, I disagree here.
>There should be only one limit as on line 1322, and there is no need for substitutions n -> oo or oo -> n

That is only for the case when number of singularities is `0` which is not the case here. Here number of singularities is 1. The singularity being at `x = 0`

>`- limit(F, x, a) + limit(F, x, b)`

What looks like the difference of limits is actually this expression :-
```
limit (  integration from "a",  breaking at singularity(ies) , finally going towards "b" , r, 0 '+' )
```
which was written in the code as this ( for `len(singularities) == 1` ) :
```
limit(((F.subs(x, singularities_list[0] - r)- limit(F, x, a).subs([(oo, n), (-oo, -n)])) + limit(F, x, b).subs([(oo, n), (-oo, -n)]) - F.subs(x, singularities_list[0] + r)), r, 0,'+')  
```
with `- limit(F, x, a)` and `+ limit(F, x, b)` written outside `limit r-->0` so as to simplify the expression since `- limit(F, x, a)` and `+ limit(F, x, b)` are independent of `r`

So I'm sure there's nothing incorrect with the way `subs` is used.

Reply:
@jksuom Sir, I have written the code exactly on the lines of the definition given [here](http://mathworld.wolfram.com/CauchyPrincipalValue.html) which includes the location where the limit is used (`limit r-->0`).
So I'm pretty sure about the working of this now.

Reply:
> That is only for the case when number of singularities is 0 which is not the case here. 

That is for all cases by the definition of principal value if the limits are `-oo` and `oo`. But note that it is only part of the definition, not the whole integral in general. Other terms coming from the singularities shall be added to obtain the final result.